keps/1224-lending-limit/README.md

```diff
@@ -0,0 +1,162 @@
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+  - [Kueue LendingLimit API](#kueue-lendinglimit-api)
+    - [Note](#note)
+  - [Test Plan](#test-plan)
+      - [Prerequisite testing updates](#prerequisite-testing-updates)
+    - [Unit Tests](#unit-tests)
+    - [Integration tests](#integration-tests)
+  - [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+  <!-- /toc -->
+
+## Summary
+
+Under the current implementation, one ClusterQueue's resources could be borrowed completely by others in the same cohort, this improves the resource utilization to some extent, but sometimes, user wants to reserve some resources only for private usage. 
+
+This proposal provides a guarantee mechanism for users to solve this problem. They can have a reservation of resource quota that will never be borrowed by other clusterqueues in the same cohort.
+
+## Motivation
+
+Sometimes we want to keep some resources for guaranteed usage, so that when new jobs come into queue, they can be admitted immediately.
+
+Under the current implementation, we are using `BorrowingLimit` to define the maximum amount of quota that this ClusterQueue is allowed to borrow. But this may cause another ClusterQueue in the same cohort to run out of resources.
+
+Even if we set the `Preemption`, it still needs some time and spends a lot of unnecessary cost.
+
+So we need a reservation design for resource requests and security reasons: `LendingLimit`, to claim the quota allowed to lend, reserve a certain amount of resources to ensure that they will never be borrowed.
+
+### Goals
+
+- Implement `LendingLimit`, users can have a reservation of guaranteed resource by claiming the `LendingLimit`.
+
+### Non-Goals
+
+- Replace `BorrowingLimit` to some extent in the future.
+
+## Proposal
+
+In this proposal, `LendingLimit` is defined. The `ClusterQueue` will be limited to lend the specified quota to other ClusterQueues in the same cohort.
+
+### User Stories (Optional)
+
+#### Story 1
+
+In order to ensure the full utilization of resources, we generally set `BorrowingLimit` to max, but this may cause a `ClusterQueue` to run out of its all resources, and make the new incoming job slow to response for the slow preemption. This could be worse in a competitive cluster, jobs will borrow resources and be reclaimed over and over.
+
+So we want to reserve some resources for a `ClusterQueue`, so that any incoming jobs in the  `ClusterQueue` can get admitted immediately.
+
+### Notes/Constraints/Caveats (Optional)
+
+With both BorrowingLimit and LendingLimit configured, one clusterQueue may not be able to borrow up to the limit just because we reserved the lending limit quota of resource.
+
+To reduce confusion, we will recommend to users to only set borrowingLimit or lendingLimit, but not both, even though both will be supported at the same time.
```

Comment:
> I'm ok either way. I remember seeing a similar recommendation in YARN not to use borrowingLimit and weight at the same time.
> 
> Another alternative would be some form of troubleshooting, for example:
> 
> > Why is my CQ not borrowing?
> > Check borrowingLimit of your CQ and lendingLimit of other CQs in the cohort.

I prefer to take this case (`BorrowingLimit` and `LendingLimit`) as a troubleshooting technique.

Reply:
Sounds reasonable. I agree to remove Line 66.

Reply:
Plz update the words.

Reply:
> This is not about we should only set `borrowingLimit` or `lendingLimit` or both, but a practice that if we configured the `lendingLimit` for demand, then `borrowingLimit` could be omitted for the maximum resource utilization as there's no borrowingLimit.

@kerthcet Does this mean the following situation?

cq-a and cq-b belong to the same cohort.

- cq-a:
   cpu: 10
   borrowingLimit: 5

- cq-b:
   cpu: 5
   lendingLimit: 0

Reply:
You can set the borrowingLimit or lendingLimit by your demands, but if we set the lendingLimit in cq-b, then cq-a can leave the borrowingLimit unset to consume the resources as much as possible, as
```
- cq-a:
  cpu: 10
  # borrowingLimit: 5 # no need to set this if you want to maximum the resource utilization 
  # and don't need to worry about use too much resources because other clusterQueues can 
  # constrain this by setting lendingLimit.

- cq-b:
  cpu: 5
  lendingLimit: X
```